README.md
# D Weather Forecast
Get weather forecast from http://forecast.weather.gov with the D programming language

# Example

```d
import weather_forecast : getForecast, WeatherData;
import std.stdio : stdout;

getForecast(delegate(WeatherData weather_data, Exception err) {
  if (! err) {
    stdout.writefln("%s, %s", weather_data.city, weather_data.region);
    stdout.writefln("%s", weather_data.summary);
    stdout.writefln("%s F", weather_data.temperature);
  }
});

/*
San Francisco, California
Mostly Sunny
76 F
*/

```

# Documentation

[https://workhorsy.github.io/d-weather-forecast/2.0.0/](https://workhorsy.github.io/d-weather-forecast/2.0.0/)

# Generate documentation

```
dub --build=docs
```

# Run unit tests

```
dub test
```

[![Dub version](https://img.shields.io/dub/v/d-weather-forecast.svg)](https://code.dlang.org/packages/d-weather-forecast)
[![Dub downloads](https://img.shields.io/dub/dt/d-weather-forecast.svg)](https://code.dlang.org/packages/d-weather-forecast)
[![License](https://img.shields.io/badge/license-BSL_1.0-blue.svg)](https://raw.githubusercontent.com/workhorsy/d-weather-forecast/master/LICENSE)

ChangeLog
06/07/2020 Release 2.0.0
	* Fixed Bug #6: Make tests run without main for changes in dmd 2.090

12/28/2017 Release 1.3.0
	* Fixed Bug #4: Don't return partial data on error

9/17/2017 Release 1.2.0
	* Fixed Bug #2: Only require BDD for running tests
	* Fixed Bug #3: If ipinfo fails no error is returned


test/main.d


import BDD;

unittest {
	import weather_forecast : getForecast, WeatherData;
	import ipinfo : httpGet;

	immutable string RESULT_IP =
	`{
		"ip": "8.8.8.8",
		"loc": "37.7749,-122.4194",
		"org": "AS15169 Google Inc.",
		"city": "Mountain View",
		"region": "California",
		"country": "US",
		"postal": "94043"
	}`;
	immutable string RESULT_WEATHER =
	`{
		"data": { "weather": [ "Hot" ] },
		"currentobservation": { "Temp" : "70" } }
	}`;

	httpGet = delegate(string url, void delegate(int status, string response) cb) {
		import std.string : startsWith;

		if (url.startsWith("https://ipinfo.io/json")) {
			cb(200, RESULT_IP);
		} else if (url.startsWith("http://forecast.weather.gov/MapClick.php?lat=")) {
			cb(200, RESULT_WEATHER);
		}
	};

	describe("weather_forecast",
		it("Should get a forecast", delegate() {
			getForecast(delegate(WeatherData weather_data, Exception err) {
				err.shouldBeNull();

				weather_data.latitude.shouldEqual("37.7749");
				weather_data.longitude.shouldEqual("-122.4194");
				weather_data.city.shouldEqual("Mountain View");
				weather_data.region.shouldEqual("California");
				weather_data.country.shouldEqual("US");
				weather_data.postal.shouldEqual("94043");
				weather_data.temperature.shouldEqual("70");
				weather_data.summary.shouldEqual("Hot");
			});
		}),
		it("Should return an error when failing to parse ipinfo json response", delegate() {
			httpGet = delegate(string url, void delegate(int status, string response) cb) {
				import std.string : startsWith;

				if (url.startsWith("https://ipinfo.io/json")) {
					cb(200, "{");
				} else if (url.startsWith("http://forecast.weather.gov/MapClick.php?lat=")) {
					cb(200, RESULT_WEATHER);
				}
			};

			getForecast(delegate(WeatherData weather_data, Exception err) {
				err.shouldNotBeNull();
				err.msg.shouldEqual(`Failed to parse "https://ipinfo.io/json" JSON response`);
			});
		}),
		it("Should return an error when the ipinfo server fails", delegate() {
			httpGet = delegate(string url, void delegate(int status, string response) cb) {
				import std.string : startsWith;

				if (url.startsWith("https://ipinfo.io/json")) {
					cb(500, "");
				} else if (url.startsWith("http://forecast.weather.gov/MapClick.php?lat=")) {
					cb(200, RESULT_WEATHER);
				}
			};

			getForecast(delegate(WeatherData weather_data, Exception err) {
				err.shouldNotBeNull();
				err.msg.shouldEqual(`Request for "https://ipinfo.io/json" failed with status code: 500`);
			});
		}),
		it("Should return an error when failing to parse weather json response", delegate() {
			httpGet = delegate(string url, void delegate(int status, string response) cb) {
				import std.string : startsWith;

				if (url.startsWith("https://ipinfo.io/json")) {
					cb(200, RESULT_IP);
				} else if (url.startsWith("http://forecast.weather.gov/MapClick.php?lat=")) {
					cb(200, "{");
				}
			};

			getForecast(delegate(WeatherData weather_data, Exception err) {
				err.shouldNotBeNull();
				err.msg.shouldEqual(`Failed to parse "http://forecast.weather.gov/MapClick.php?lat=37.7749&lon=-122.4194&FcstType=json" JSON response`);
			});
		}),
		it("Should return an error when the weather server fails", delegate() {
			httpGet = delegate(string url, void delegate(int status, string response) cb) {
				import std.string : startsWith;

				if (url.startsWith("https://ipinfo.io/json")) {
					cb(200, RESULT_IP);
				} else if (url.startsWith("http://forecast.weather.gov/MapClick.php?lat=")) {
					cb(500, "");
				}
			};

			getForecast(delegate(WeatherData weather_data, Exception err) {
				err.shouldNotBeNull();
				err.msg.shouldEqual(`Request for "http://forecast.weather.gov/MapClick.php?lat=37.7749&lon=-122.4194&FcstType=json" failed with status code: 500`);
			});
		}),
	);
}

int main() {
	return BDD.printResults();
}


test/test.d



unittest {
	import weather_forecast : getForecast, WeatherData;
	import ipinfo : httpGet;
	import BDD;

	immutable string RESULT_IP =
	`{
		"ip": "8.8.8.8",
		"loc": "37.7749,-122.4194",
		"org": "AS15169 Google Inc.",
		"city": "Mountain View",
		"region": "California",
		"country": "US",
		"postal": "94043"
	}`;
	immutable string RESULT_WEATHER =
	`{
		"data": { "weather": [ "Hot" ] },
		"currentobservation": { "Temp" : "70" } }
	}`;

	httpGet = delegate(string url, void delegate(int status, string response) cb) {
		import std.string : startsWith;

		if (url.startsWith("https://ipinfo.io/json")) {
			cb(200, RESULT_IP);
		} else if (url.startsWith("http://forecast.weather.gov/MapClick.php?lat=")) {
			cb(200, RESULT_WEATHER);
		}
	};

	describe("weather_forecast",
		it("Should get a forecast", delegate() {
			getForecast(delegate(WeatherData weather_data, Exception err) {
				err.shouldBeNull();

				weather_data.latitude.shouldEqual("37.7749");
				weather_data.longitude.shouldEqual("-122.4194");
				weather_data.city.shouldEqual("Mountain View");
				weather_data.region.shouldEqual("California");
				weather_data.country.shouldEqual("US");
				weather_data.postal.shouldEqual("94043");
				weather_data.temperature.shouldEqual("70");
				weather_data.summary.shouldEqual("Hot");
			});
		}),
		it("Should return an error when failing to parse ipinfo json response", delegate() {
			httpGet = delegate(string url, void delegate(int status, string response) cb) {
				import std.string : startsWith;

				if (url.startsWith("https://ipinfo.io/json")) {
					cb(200, "{");
				} else if (url.startsWith("http://forecast.weather.gov/MapClick.php?lat=")) {
					cb(200, RESULT_WEATHER);
				}
			};

			getForecast(delegate(WeatherData weather_data, Exception err) {
				err.shouldNotBeNull();
				err.msg.shouldEqual(`Failed to parse "https://ipinfo.io/json" JSON response`);
			});
		}),
		it("Should return an error when the ipinfo server fails", delegate() {
			httpGet = delegate(string url, void delegate(int status, string response) cb) {
				import std.string : startsWith;

				if (url.startsWith("https://ipinfo.io/json")) {
					cb(500, "");
				} else if (url.startsWith("http://forecast.weather.gov/MapClick.php?lat=")) {
					cb(200, RESULT_WEATHER);
				}
			};

			getForecast(delegate(WeatherData weather_data, Exception err) {
				err.shouldNotBeNull();
				err.msg.shouldEqual(`Request for "https://ipinfo.io/json" failed with status code: 500`);
			});
		}),
		it("Should return an error when failing to parse weather json response", delegate() {
			httpGet = delegate(string url, void delegate(int status, string response) cb) {
				import std.string : startsWith;

				if (url.startsWith("https://ipinfo.io/json")) {
					cb(200, RESULT_IP);
				} else if (url.startsWith("http://forecast.weather.gov/MapClick.php?lat=")) {
					cb(200, "{");
				}
			};

			getForecast(delegate(WeatherData weather_data, Exception err) {
				err.shouldNotBeNull();
				err.msg.shouldEqual(`Failed to parse "http://forecast.weather.gov/MapClick.php?lat=37.7749&lon=-122.4194&FcstType=json" JSON response`);
			});
		}),
		it("Should return an error when the weather server fails", delegate() {
			httpGet = delegate(string url, void delegate(int status, string response) cb) {
				import std.string : startsWith;

				if (url.startsWith("https://ipinfo.io/json")) {
					cb(200, RESULT_IP);
				} else if (url.startsWith("http://forecast.weather.gov/MapClick.php?lat=")) {
					cb(500, "");
				}
			};

			getForecast(delegate(WeatherData weather_data, Exception err) {
				err.shouldNotBeNull();
				err.msg.shouldEqual(`Request for "http://forecast.weather.gov/MapClick.php?lat=37.7749&lon=-122.4194&FcstType=json" failed with status code: 500`);
			});
		}),
	);
}


source/weather_forecast.d
// Copyright (c) 2017-2020 Matthew Brennan Jones <matthew.brennan.jones@gmail.com>
// Boost Software License - Version 1.0
// Get weather forecast with the D programming language
// https://github.com/workhorsy/d-weather-forecast

/++
Get weather forecast with the D programming language. It first gets your longitude
and latitude using http://ipinfo.io. Then uses them to look up your
weather using http://forecast.weather.gov.

Home page:
$(LINK https://github.com/workhorsy/d-weather-forecast)

Version: 2.0.0

License:
Boost Software License - Version 1.0

Examples:
----
import std.stdio : stdout, stderr;
import weather_forecast : getForecast, WeatherData;

getForecast(delegate(WeatherData weather_data, Exception err) {
	if (err) {
		stderr.writefln("%s", err);
	} else {
		stdout.writefln("latitude: %s", weather_data.latitude);
		stdout.writefln("longitude: %s", weather_data.longitude);
		stdout.writefln("city: %s", weather_data.city);
		stdout.writefln("region: %s", weather_data.region);
		stdout.writefln("country: %s", weather_data.country);
		stdout.writefln("postal: %s", weather_data.postal);
		stdout.writefln("temperature: %s", weather_data.temperature);
		stdout.writefln("summary: %s", weather_data.summary);
	}
});
----
+/

module weather_forecast;


/++
Data gathered in WeatherData:
----
struct WeatherData {
	string latitude;
	string longitude;
	string city;
	string region;
	string country;
	string postal;
	string temperature;
	string summary;
}
----
+/

struct WeatherData {
	string latitude;
	string longitude;
	string city;
	string region;
	string country;
	string postal;
	string temperature;
	string summary;
}

/++
Returns the weather forecast using a callback.

Params:
 cb = The callback to fire when weather info has been downloaded. The callback
	sends the WeatherData data and an Exception if there was any problem.
+/
void getForecast(void delegate(WeatherData weather_data, Exception err) cb) {
	import std.stdio : stdout, stderr;
	import std.json : JSONValue, parseJSON;
	import std.string : chomp, format;
	import std.array : split;
	import std.conv : to;
	import ipinfo : getIpinfo, IpinfoData, httpGet;

	WeatherData weather_data;

	getIpinfo(delegate(IpinfoData ipinfo_data, Exception err) {
		if (err) {
			cb(WeatherData.init, err);
		} else {
			const string URL = "http://forecast.weather.gov/MapClick.php?lat=" ~ ipinfo_data.latitude ~ "&lon=" ~ ipinfo_data.longitude ~ "&FcstType=json";

			httpGet(URL, delegate(int status, string response) {
				if (status != 200) {
					auto err = new Exception("Request for \"%s\" failed with status code: %s".format(URL, status));
					cb(WeatherData.init, err);
					return;
				}

				try {
					JSONValue j = parseJSON(response);

					weather_data.latitude = ipinfo_data.latitude;
					weather_data.longitude = ipinfo_data.longitude;
					weather_data.city = ipinfo_data.city;
					weather_data.region = ipinfo_data.region;
					weather_data.country = ipinfo_data.country;
					weather_data.postal = ipinfo_data.postal;
					weather_data.temperature = j["currentobservation"]["Temp"].str();
					weather_data.summary = j["data"]["weather"][0].str();
				} catch (Throwable) {
					auto err = new Exception("Failed to parse \"%s\" JSON response".format(URL));
					cb(WeatherData.init, err);
					return;
				}

				cb(weather_data, null);
			});
		}
	});
}


source/weather.d
// Copyright (c) 2017 Matthew Brennan Jones <matthew.brennan.jones@gmail.com>
// Boost Software License - Version 1.0
// Get weather forecast with the D programming language
// https://github.com/workhorsy/d-weather-forecast

/++
Get weather forecast with the D programming language. It first gets your longitude
and latitude using http://ipinfo.io. Then looks up your weather using
http://forecast.weather.gov.

Home page:
$(LINK https://github.com/workhorsy/d-weather-forecast)

Version: 1.0.0

License:
Boost Software License - Version 1.0

Examples:
----
import std.stdio : stdout;
import WeatherForecast : getForecast;

getForecast(delegate(string latitude, string longitude, string temperature, string weather) {
	stdout.writefln("latitude: %s", latitude);
	stdout.writefln("longitude: %s", longitude);
	stdout.writefln("temperature: %s", temperature);
	stdout.writefln("weather: %s", weather);
});
----
+/

module WeatherForecast;

void delegate(string url, void delegate(int status, string response) cb) httpGet;
private void delegate(string url, void delegate(int status, string response) cb) httpGetDefault;

static this() {
	httpGetDefault = delegate(string url, void delegate(int status, string response) cb) {
		import std.stdio : stdout, stderr;
		import std.net.curl : HTTP, CurlException, get;

		auto http = HTTP();
		string content = "";

		try {
			content = cast(string) get(url, http);
		} catch (CurlException ex) {
			stderr.writefln("!!! url: %s", url);
			stderr.writefln("!!! CurlException: %s", ex.msg);
			//stderr.writefln("!!!!!!!!!!!!!!!! CurlException: %s", ex);
		}

		ushort status = http.statusLine().code;
		cb(status, content);
	};

	httpGet = httpGetDefault;
}

/++
Returns the weather forecast using a callback.

Params:
 cb = The callback to fire when weather info has been downloaded.
+/
void getForecast(void delegate(string latitude, string longitude, string temperature, string weather) cb) {
	import std.stdio : stdout, stderr;
	import std.json : JSONValue, parseJSON;
	import std.string : chomp;
	import std.array : split;
	import std.conv : to;

	// Get latitude and longitude from ip address
	httpGet("http://ipinfo.io/loc", delegate(int status, string response) {
		if (status != 200) {
			stderr.writefln("Request for lat and lon data failed with status code: %s", status);
			return;
		}

		string[] result = split(response, ",");
		string latitude = chomp(result[0]);
		string longitude = chomp(result[1]);

		string url = "http://forecast.weather.gov/MapClick.php?lat=" ~ latitude ~ "&lon=" ~ longitude ~ "&FcstType=json";
		httpGet(url, delegate(int status, string response) {
			if (status != 200) {
				stderr.writefln("Request for Weather data failed with status code: %s", status);
				return;
			}

			string temperature = "";
			string weather = "";
			try {
				JSONValue j = parseJSON(response);
				temperature = j["currentobservation"]["Temp"].str();
				weather = j["data"]["weather"][0].str();
			} catch (Throwable) {
				stderr.writefln("Failed to parse Weather server JSON response: %s", response);
				return;
			}

			cb(latitude, longitude, temperature, weather);
		});
	});
}

unittest {
	import BDD;

	immutable string RESULT_IP = "37.7749,-122.4194\r\n";
	immutable string RESULT_WEATHER =
	`{
		"data": { "weather": [ "Hot" ] },
		"currentobservation": { "Temp" : "70" } }
	}`;

	WeatherForecast.httpGet = delegate(string url, void delegate(int status, string response) cb) {
		import std.string : startsWith;

		if (url.startsWith("http://ipinfo.io/loc")) {
			cb(200, RESULT_IP);
		} else if (url.startsWith("http://forecast.weather.gov/MapClick.php?lat=")) {
			cb(200, RESULT_WEATHER);
		}
	};

	describe("WeatherForecast",
		it("Should get a forecast", delegate() {
			WeatherForecast.getForecast(delegate(string latitude, string longitude, string temperature, string weather) {
				latitude.shouldEqual("37.7749");
				longitude.shouldEqual("-122.4194");
				temperature.shouldEqual("70");
				weather.shouldEqual("Hot");
			});
		}),
	);
}
